src/nexuscli/nexus_client.py

```diff
@@ -483,3 +498,163 @@ def upload(self, source, destination):
+        """False when nocache is set or local file is out-of-date"""
+        if nocache:
+            try:
+                LOG.debug('Removing {} because nocache is set\n'.format(
+                    download_path))
+                os.remove(download_path)
+            except FileNotFoundError:
+                pass
+            return False
+
+        for hash_name in ['sha1', 'md5']:
+            h = artefact.get('checksum', {}).get(hash_name)
+            if h is None:
+                continue
+
+            if self._local_hash_matches_remote(download_path, h, hash_name):
+                LOG.debug('Skipping {download_url} because local copy '
+                          '{download_path} is up-to-date\n'.format(**locals()))
+                return True
+
+        return False
+
+    def download_file(self, download_url, destination):
+        """Download an asset from Nexus artefact repository to local
+        file system.
+
+        :param download_url: fully-qualified URL to asset being downloaded.
+        :param destination: file or directory location to save downloaded
+            asset. Must be an existing directory; any exiting file in this
+            location will be overwritten.
+        :return:
+        """
+        response = self._get(download_url)
+
+        if response.status_code != 200:
+            sys.stderr.write(response.__dict__)
+            raise DownloadError(
+                'Downloading from {download_url}. '
+                'Reason: {response.reason}'.format(**locals()))
+
+        with open(destination, 'wb') as fd:
+            LOG.debug('Writing {download_url} to {destination}\n'.format(
+                **locals()))
+            for chunk in response.iter_content():
+                fd.write(chunk)
+
+    def download(self, source, destination, log_lvl='WARNING',
+                 flatten=False, nocache=False):
+        """Process a download. The source must be a valid Nexus 3
+        repository path, including the repository name as the first component
+        of the path.
+
+        The destination must be a local file name or directory.
+
+        If a file name is given as destination, the asset may be renamed. The
+        final destination will depend on self.flatten: when True, the remote
+        path isn't reproduced locally.
+
+        :param source: location of artefact or directory on the repository
+            service.
+        :param destination: path to the local file or directory.
+        :param flatten: when True, the remote path isn't reproduced locally.
+        :param nocache: Force download of a directory or artefact even if local
+                        copy is available and is up-to-date with the version
+                        available on Nexus.
+        :return: number of downloaded files.
+        """
+        # FIXME: Enable logging at application level rather than in this method
+        LOG.setLevel(level=log_lvl)
```

Comment:
This presentation from PyCon au 2018 (by @funkybob) might be useful. The "ah-ha" moment for me was that log levels can be set both at the logger and handler objects and the lowest wins (ie. if you set one to debug and the other to warning, you'll only see warning and above).

https://www.youtube.com/watch?v=DxZ5WEo4hvU&feature=youtu.be&t=4m20s